docs/syntax-table-destructuring.md

```diff
@@ -0,0 +1,143 @@
+# Table Destructuring Syntax
+
+
+
+## Summary
+Introduce new syntax for unpacking values from tables.
+
+```lua
+local .{ .foo, one, ["@bar"] as bar } = thing
+-- foo == thing.foo
+-- one == thing[1]
+-- bar == thing["@bar"]
+```
+
+This RFC addresses issues brought up from previous table destructuring RFC's.
+
+
+
+## Motivation
+Aliasing table items through variable assignments is a common pattern in luau, especially when working with large libraries. However doing this with a lot of items becomes repetitve:
+```lua
+local useEffect = React.useEffect
+local useMemo = React.useMemo
+local useState = React.useState
+```
+
+With destructuring syntax this could be shortened to the following:
+```lua
+local .{ .useEffect, .useMemo, .useState } = React
+```
+
+## Design
+
+We propose that the table destructuring syntax is prefixed with a period (`.`) symbol. Without a prefix destructuring becomes ambiguous in certain scenarios, such as the one below.
+```lua
+baz = foo
+{ bar } = test
+```
+*(This is ambiguous as it can be interpretted as `baz = foo { bar }; = test` or `baz = foo; { bar } = test`).*
+
+### Key destructuring
+We propose the following syntax for destructuring by keys:
+```lua
+.{ ["foo"] as foo } = thing
+-- foo == thing["foo"]
+```
+
+If the key can be expressed as a valid variable name then a shorthand can be used instead:
+```lua
+.{ .foo } = thing
+```
+
+Shorthands can also be destructured with a different name:
+```lua
+.{ .foo as bar } = thing
+-- bar == thing.foo
+```
+
+We propose nested key destructuring via the following syntax (This also works with shorthands):
+```lua
+.{ ["@foo"] as .{ .bar } } = thing
+-- bar == thing["@foo"].bar
+```
+
+## Array destructuring
+We propose the following syntax for array destructuring:
+```lua
+.{ one, two, three } = thing
+-- one == thing[1]
+-- two == thing[2]
+-- three == thing[3]
+```
+
+We propose the following syntax for nested array destructuring:
+```lua
+.{ one, .{ apple }, three } = thing
+-- one == thing[1]
+-- apple == thing[2][1]
+-- three == thing[3]
+```
+
+## Mixed Destructuring
+Key and array destructuring can be used together:
+```lua
+.{ .foo, one, .bar as .{ .baz }, .{ apple } } = thing
+-- foo == thing.foo
+-- one == thing[1]
+-- baz == thing.bar.baz
+-- apple = thing[2][1]
+```
+
+## Rest Destructuring
+
+We propose destructuring the rest of the properties via the following syntax (`rest` can be any identifier). Only one rest parameter can be defined.
+```lua
+.{ ["@foo"] as .{ .bar }, ...rest } = thing
+```
```

Comment:
Please explain the effect of this statement and how it will be implemented.

Reply:
Doesn't look implementable to me. How can this work with `userdata`? You can't iterate over keys in `__index`, or remove a key from `__index`, not to mention that `table.clone` doesn't work if the table's metatable has `__metatable`, so it follows that `local { ...clone } = t` must follow the same rule as `table.clone`.